File: leetcode/problems/problem-specific/56.merge-intervals/56.merge-intervals.go
package main

import "sort"

func merge(intervals [][]int) [][]int {
	if len(intervals) <= 1 {
		return intervals
	}

	res, n, i := [][]int{}, len(intervals), 0
	// * Sort the intervals based on start value
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	for i = 0; i < n-1; i++ {
		start, end := intervals[i][0], intervals[i][1]
		for i < n-1 && end >= intervals[i+1][0] {
			end = findMax(end, intervals[i+1][1])
			i++
		}

		res = append(res, []int{start, end})
	}

	if i == n-1 {
		res = append(res, intervals[i])
	}

	return res
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

File: leetcode/problems/language-specific/python/56.merge-intervals.py
#
# @lc app=leetcode id=56 lang=python3
#
# [56] Merge Intervals
#

# @lc code=start

# * Sorting Solution | O(nlogn) Time | O(n) Space
# * n -> The length of intervals array


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        # * Sort the intervals in ascending order based on the start value.
        intervals.sort(key=lambda interval: interval[0])
        merged_intervals = []
        for interval in intervals:
            # * Add the current interval to the merged intervals if the merged intervals is empty
            # * or if the current interval doesn't overlap with the previous merged interval.
            if not merged_intervals or merged_intervals[-1][1] < interval[0]:
                merged_intervals.append(interval)
            else:
                # * Else there is overlap, so we merge the current and previous intervals.
                merged_intervals[-1][1] = max(merged_intervals[-1][1], interval[1])

        return merged_intervals


# @lc code=end
